File: leetcode/543.二叉树的直径/543.二叉树的直径.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func diameterOfBinaryTree(root *TreeNode) int {
	ans := 0
	var depth func(root *TreeNode) int
	depth = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		left := depth(root.Left)
		right := depth(root.Right)
		if left+right > ans {
			ans = left + right
		}
		if left > right {
			return left + 1
		} else {
			return right + 1
		}
	}
	depth(root)
	return ans
}

File: leetcode/543.二叉树的直径/543.二叉树的直径.py
from util.leetcode_type import TreeNode


class Solution:
    def __init__(self):
        self.ans = 0

    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        """
        二叉树的直径：从一个节点到另一个节点最长的路径
        题目求：任意节点，所以要每次递归是都判断当前的路径是否是最大
        """

        def deep(root):
            if root is None:
                return 0
            left = deep(root.left)
            right = deep(root.right)
            # 判断此节点路径是否是最长
            self.ans = max(self.ans, left + right + 1)
            # 返回左右路径中最长+1
            return max(left, right)

        deep(root)
        return self.ans
